### PR TITLE
Add per-project hashing salt

### LINKED ISSUE #343: Per project salt in PIXL hashing
### Definition of Done / Acceptance Criteria

- [ ] PIXL hasher uses the configured project yaml files to have a project-specific salt

### Testing

_No response_

### Documentation

_No response_

### Dependencies

- Anonymisation library work

### Details and Comments

We're now keeping the PIXL-specific hashing API for this iteration so need to make sure that we use a different salt per project in hashing

Suggested flow

- [ ] Use project name as the key in azure keyvault, which stores the salt (note: separate Azure Keyvault than the one used for project destination secrets)
- [ ] If the key doesn't exist, create it and save it in the keyvault
- [ ] in memory cache the salt values (in the hashing api)
- [ ] Use this salt in secure hashing

### PR DESCRIPTION
**Pull request by milanmlft**
Fixes #343 

* Use `core.project_config.secrets.AzureKeyVault` in `hasher` to set up Azure Key Vault connection
* Hashing salts are now retrieved from the key vault, using the project-name as key
	* If no salt exists, a new one is created and stored in the key vault
* Removed the old `hasher` key vault functionality
* Salt generation is moved into the hashing functionality, so only `hasher` knows about it

### To do

- [x] Wait until #366 gets merged into `main` before continuing

### CODE REVIEW

File: pytest-pixl/src/pytest_pixl/plugin.py
Comment (stefpiatek) [inline pytest-pixl/src/pytest_pixl/plugin.py:82]: is this fixture used anywhere?

File: pytest-pixl/src/pytest_pixl/keyvault.py
Comment (stefpiatek) [inline pytest-pixl/src/pytest_pixl/keyvault.py:17]: feels like this should just belong in hasher api tests if we're only using it there? We can then remove pytest pixl from depdendencies

File: pixl_core/src/core/project_config/secrets.py
Comment (stefpiatek) [inline pixl_core/src/core/project_config/secrets.py:109]: all arguments need to be hashable for caching to be enabled. So you could make SecretClient hashable or keep this as before where it was being cached

File: pixl_core/src/core/project_config/secrets.py
Comment (stefpiatek) [inline pixl_core/src/core/project_config/secrets.py:87]: see below for all arguments having to be hashable. check locally that the caching is truly being used

File: hasher/src/hasher/endpoints.py
Comment (stefpiatek) [inline hasher/src/hasher/endpoints.py:36]: could we just have a `hasher` object created upon startup of the app and we use that in all of the subcalls?

File: .github/workflows/main.yml
Comment (stefpiatek) [inline .github/workflows/main.yml:63]: Probabyl could strip out pytest-pixl?

File: hasher/src/hasher/hashing.py
Comment (stefpiatek) [inline hasher/src/hasher/hashing.py:39]:
```suggestion
    Hasher class to generate keyed hash digests using the Blake2b algorith. 
    Gets or creates the salt for a specific project from the Azure Key Vault instance.
```

File: hasher/src/hasher/hashing.py
Comment (stefpiatek) [inline hasher/src/hasher/hashing.py:76]: maybe `get_or_create_salt`?

**Review by stefpiatek — COMMENTED** [review]
Looking nice, but refactoring of caching won't actually cache sadly

File: hasher/src/hasher/endpoints.py
Comment (milanmlft) [reply hasher/src/hasher/endpoints.py:36]: Yeah so I had that originally, but that's where I ran into the `pytest` setup problems, because it would create that `hasher` object during one of the imports, before my monkeypatch to use a mock key vault was applied, so it would try to set up an actual key vault connection with `AzureKeyVault()`, which would then fail because the environment variables are not set... I got really confused as to why the monkeypatch wasn't working as I expected, so ended up just moving these calls inside API calls 🤷‍♂️ 

But agreed that it's not very nice. On the other hand, I'm now actually in the process of removing all the other endpoints except for `/hash`, and moving the special handling logic for patient ID and accession number into `dcmd`, which feels more reasonable?

I'll have another look to see if I can fix the hasher object creation in a way that is compatible with the tests (or vice versa).

File: pytest-pixl/src/pytest_pixl/plugin.py
Comment (milanmlft) [reply pytest-pixl/src/pytest_pixl/plugin.py:82]: Ah yeah good point, this is leftover from an earlier attempt

File: pytest-pixl/src/pytest_pixl/keyvault.py
Comment (milanmlft) [reply pytest-pixl/src/pytest_pixl/keyvault.py:17]: Yeah was originally thinking of also using it for the `core.uploader` tests, which is why I put it here, but maybe not that useful.

File: hasher/src/hasher/endpoints.py
Comment (stefpiatek) [reply hasher/src/hasher/endpoints.py:36]: what are we importing in the `conftest` which ends up importing this module? Should be able to monkeypatch if we're not importing it, and I wouldn't expect the conftest to import much, but could always import locally in a fixture to avoid this?

Comment (milanmlft) [reply hasher/src/hasher/endpoints.py]: I updated the `Hasher` class slightly. Now it takes the project slug as a parameter during initialisation, so now it has to be created within the API call anyway, I think?

Comment (milanmlft) [reply hasher/src/hasher/endpoints.py]: Resolving for now, but open to more feedback!

**Comment by milanmlft** [comment]
`pixl_dcmd` tests failing but will wait until #366 gets merged in before resolving.

**Comment by milanmlft** [comment]
Docs could definitely still use some further improvement.

**Review by stefpiatek — APPROVED** [review]
Nice! I think some env changes are needed and a couple of questions about the public API

File: hasher/pyproject.toml
Comment (milanmlft) [reply hasher/pyproject.toml:13]: Aah this is a legacy dependency, not used anymore since I deleted `hasher.settings`. Will remove!

File: hasher/src/hasher/endpoints.py
Comment (milanmlft) [reply hasher/src/hasher/endpoints.py:41]: Yeah good point, this was already feeling a bit awkward. So would we then still want a separate `/salt` endpoint for that?

File: hasher/src/hasher/endpoints.py
Comment (milanmlft) [reply hasher/src/hasher/endpoints.py:40]: Not sure, this was mainly taken over from the previous implementation. Though I was wondering whether this was actually ever used.

File: pixl_core/src/core/project_config/secrets.py
Comment (milanmlft) [reply pixl_core/src/core/project_config/secrets.py:113]: Ah yes, probably doesn't make sense to cache then.

**Comment by milanmlft** [comment]
Had to create a new secret in our dev Azure Keyvault under `hasher--key` to get system test passing. The Hasher API reads the secret configured with `HASHER_API_AZ_KEY_VAULT_SECRET_NAME` from the vault as key for the hashing. Added this to `test/.env` as well.

**Pull request merged**
